src/daft-writers/src/partition.rs

```diff
@@ -0,0 +1,148 @@
+use std::{
+    collections::{hash_map::RawEntryMut, HashMap},
+    sync::Arc,
+};
+
+use common_error::DaftResult;
+use daft_core::{array::ops::as_arrow::AsArrow, utils::identity_hash_set::IndexHash};
+use daft_dsl::ExprRef;
+use daft_io::IOStatsContext;
+use daft_micropartition::MicroPartition;
+use daft_table::Table;
+
+use crate::{FileWriter, WriterFactory};
+
+/// PartitionedWriter is a writer that partitions the input data by a set of columns, and writes each partition
+/// to a separate file. It uses a map to keep track of the writers for each partition.
+struct PartitionedWriter {
+    per_partition_writers:
+        HashMap<IndexHash, Box<dyn FileWriter<Input = Arc<MicroPartition>, Result = Vec<Table>>>>,
+    saved_partition_values: Vec<Table>,
+    writer_factory: Arc<dyn WriterFactory<Input = Arc<MicroPartition>, Result = Vec<Table>>>,
+    partition_by: Vec<ExprRef>,
+}
+
+impl PartitionedWriter {
+    pub fn new(
+        writer_factory: Arc<dyn WriterFactory<Input = Arc<MicroPartition>, Result = Vec<Table>>>,
+        partition_by: Vec<ExprRef>,
+    ) -> Self {
+        Self {
+            per_partition_writers: HashMap::new(),
+            saved_partition_values: vec![],
+            writer_factory,
+            partition_by,
+        }
+    }
+
+    fn partition(
+        partition_cols: &[ExprRef],
+        data: &Arc<MicroPartition>,
+    ) -> DaftResult<(Vec<Table>, Table)> {
+        let data = data.concat_or_get(IOStatsContext::new("MicroPartition::partition_by_value"))?;
+        let table = data.first().unwrap();
+        let (split_tables, partition_values) = table.partition_by_value(partition_cols)?;
+        Ok((split_tables, partition_values))
+    }
+}
+
+impl FileWriter for PartitionedWriter {
+    type Input = Arc<MicroPartition>;
+    type Result = Vec<Table>;
+
+    fn write(&mut self, input: &Arc<MicroPartition>) -> DaftResult<()> {
+        let (split_tables, partition_values) =
+            Self::partition(self.partition_by.as_slice(), input)?;
+        let partition_values_hash = partition_values.hash_rows()?;
+        for (idx, (table, partition_value_hash)) in split_tables
+            .into_iter()
+            .zip(partition_values_hash.as_arrow().values_iter())
+            .enumerate()
+        {
+            let partition_value_row = partition_values.slice(idx, idx + 1)?;
+            let entry = self.per_partition_writers.raw_entry_mut().from_hash(
+                *partition_value_hash,
+                |other| {
+                    (*partition_value_hash == other.hash) && {
+                        let other_table =
+                            self.saved_partition_values.get(other.idx as usize).unwrap();
+                        other_table == &partition_value_row
+                    }
+                },
+            );
+            match entry {
```

Comment:
not a fan of using the raw pattern outside of the probe table.

Not blocking it for this PR but we should have some kind of way to create Scalars so we can just cram those into the hashmap

Reply:
Would storing the string representation of the partition values be a better alternative for now?

Reply:
Hmmm but then we'd have to do string conversion every time, so probably not.

Reply:
Let's make an issue for this and tackle this after we create `ScalarValue`